flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBaseTest.java

```diff
@@ -0,0 +1,352 @@
+                        public void go() throws Exception {
+                            testCassandraOutputFormat.close();
+                            latch.countDown();
+                        }
+                    };
+            t.start();
+            while (t.getState() != Thread.State.TIMED_WAITING) {
+                Thread.sleep(5);
+            }
+
+            Assert.assertEquals(1, testCassandraOutputFormat.getAcquiredPermits());
+            // start writing
+            completableFuture.complete(null);
+            latch.await();
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testReleaseOnSuccess() throws Exception {
+        try (TestCassandraOutputFormat openedTestCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAcquiredPermits());
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            openedTestCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+            openedTestCassandraOutputFormat.writeRecord("N/A");
+
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAcquiredPermits());
+
+            // start writing
+            completableFuture.complete(null);
+
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAcquiredPermits());
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testReleaseOnFailure() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            Assert.assertEquals(1, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            testCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+            testCassandraOutputFormat.writeRecord("N/A");
+
+            Assert.assertEquals(0, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(1, testCassandraOutputFormat.getAcquiredPermits());
+
+            completableFuture.completeExceptionally(new RuntimeException());
+
+            Assert.assertEquals(1, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+        } catch (IOException ignored) {
+            // format.close() throws the exception gathered in format.writeRecord()
+        }
```

Comment:
This is too broad. Let's explicitly call close and only guard that call with the try-catch block.

Reply:
Agree, I was too quick, I catch too much here.